### PR TITLE
Fix replication error

### PR DESCRIPTION
**Pull request by kosssi**

### CODE REVIEW

File: server/controllers/replication.coffee [hunk 1]
```diff
@@ -114,7 +114,7 @@ module.exports.proxy = (req, res, next) ->
             if req.route.path is '/replication/:id([^_]*)/:name*'
                 permissions = true
             response.on 'data', (chunk) ->
-                if req.method is 'GET'
+                if req.method is 'GET' and chunk in not [10]
```
Comment (nono) [inline server/controllers/replication.coffee]: You should add a comment to explain that this a blank line with just a `\n`.

File: server/controllers/replication.coffee [hunk 2]
```diff
@@ -114,7 +114,10 @@ module.exports.proxy = (req, res, next) ->
             if req.route.path is '/replication/:id([^_]*)/:name*'
                 permissions = true
             response.on 'data', (chunk) ->
-                if req.method is 'GET'
+
+                # chunk is [10] when server ping to allow not kill connection
+                log.info chunk.toString()
```
Comment (nono) [inline server/controllers/replication.coffee]: I think you should remove this log

**Comment by nono** [comment]
Looks good to me.

@aenario what do you think of it?

**Comment by jacquarg** [comment]
I am not sure this is the best way to fix this.
The goal of this algorithm is to test authorizations on the out stream. Data are bufferd up to the point where a document can be build, and stream :  is to get enought chunks to parse the document, and check the authorization on it. That why there were no ` catch `  bloc prior to  PR : https://github.com/cozy/cozy-data-system/pull/188 ( https://github.com/kosssi/cozy-data-system /blob/7d0ec73b603bcb908ea17c957133fb212e0d11ec/server/controllers/replication.coffee#L126-L127 )

This change shows that this algorithm doesn't fit with live replication, but, the current solution is weird because it may corrupt the out message (send '\n' before the others parts of the message). 
It might exist other solution for live replication bug (cf : https://github.com/medic/medic-webapp/issues/1503 ; and PouchDB 5.2 )

 @poupotte what your opinion around this part ?

**Comment by nono** [comment]
Okay, if I understand correctly, you suggest to not merge this PR, remove the catch block and, in cozy-mobile, update pouchdb and set up to not use timeout. Right?

On the long term, the right way to check permissions would be to have one couchdb database per doctype/permission. WIth that, we can check the permission on the request and not having to parse the response. Of course, it's not a light change, so it won't come soon.

**Comment by jacquarg** [comment]
Yep ! But, we have to confirm first that changes in cozy-mobile will work ! I will have a look right now. Thanks

**Comment by jacquarg** [comment]
Looks working : https://github.com/cozy/cozy-mobile/pull/152 . Just one ` [2016-01-22 09:52:37:246] error - replication | {} ' per live replication request (ie per 30"), removing catch sounds safe.

**Comment by aenario** [comment]
Good to merge for me. 

However, I feel this behaviour would be more cleanly represented by a `through` stream.

**Comment by kosssi** [comment]
I think it's a good PR because heartbeat can be use by another user and I think it's a good parameter for have less transfer data ;)

**Comment by nono** [comment]
I don't think it's a good idea to block all the heartbeats (`\n`). If CouchDB does send them, it's for a reason (preventing firewalls to drop the connection, I guess).

**Comment by jacquarg** [comment]
Yes, but the actual code corrupt the response :

**Comment by nono** [comment]
When the DS receive an heartbeat (`\n`), it should send it, not buffering it in `data`. Right?

**Comment by jacquarg** [comment]
ok for the heartbeat, but any ` \n ` will go to the client by passing the buffering (and so messing up the order of the data.

File: server/controllers/replication.coffee [hunk 3]
```diff
@@ -114,7 +114,10 @@ module.exports.proxy = (req, res, next) ->
             if req.route.path is '/replication/:id([^_]*)/:name*'
                 permissions = true
             response.on 'data', (chunk) ->
-                if req.method is 'GET'
+
+                # When replication are heartbeat it's send \n every 10 secondes
+                # so we don't want to have a processing
+                if req.method is 'GET' and chunk.toString() isnt "\n"
```
Comment (jacquarg) [inline server/controllers/replication.coffee]: We may just check for heartbeat in the params of the request, or something similar ?

**Comment by nono** [comment]
I don't think you will have a chunk that is only one char that is not the heartbeat.

**Comment by jacquarg** [comment]
Yes, you should be right, but it's sad to put hacks in the data-system .

**Comment by kosssi** [comment]
It's better like this ?

**Comment by jacquarg** [comment]
Yep, it's more explicit. Thanks !

**Pull request merged**